wincert.go

```diff
@@ -273,22 +284,28 @@ func buildCertChains(trusted *x509.Certificate, untrusted []*x509.Certificate, c
 	current := chain[len(chain)-1]
 
 	if !isSelfSignedCert(current) {
+		// for certificates that aren't self-signed:
+		// check the list of untrusted certs first
 		for _, x := range untrusted {
-			if alreadyInChain(x) {
-				continue
-			}
 			if !certLikelyIssued(x, current) {
 				continue
 			}
-			chains = append(chains, buildCertChains(trusted, untrusted, append(chain, x), depth)...)
+			// try to build chains with this untrusted cert
+			chains = append(chains, buildCertChains(trusted, removeCert(untrusted, x), append(chain, x), depth)...)
 		}
 
-		if !alreadyInChain(trusted) && certLikelyIssued(trusted, current) {
+		// check the trust anchor
+		if certLikelyIssued(trusted, current) {
+			// we have a complete chain
 			chains = append(chains, append(chain, trusted))
 		}
 	}
 
-	if current.Equal(trusted) {
+	// If we have no chains, check if the current certificate is the
+	// trust anchor. This handles the case where the leaf certificate
+	// is the trust anchor. We should only reach this condition at
+	// depth==1.
+	if len(chains) == 0 && certsMatch(trusted, current) {
```

Comment:
What is the purpose of checking `len(chains) == 0` here? Is that optimization? Or there are actual cases where `certsMatch(trusted, current)` and `chains` would already contain any chain? And in this case, what is the meaning of not adding the current chain?

Reply:
If it is optimization.  Maybe the comment should say so. If it is behavior, maybe we need to cover it in the tests.

Reply:
Yes, it's more of an optimization - if the leaf certificate is not the same as the trust anchor, and there is a trust relationship between them then `len(chains)` here will always be 0 and `certsMatch(trusted, current)` will always return false. I did try splitting this into an inner / outer function as well because these conditions will only ever be true on the outermost call (hence the reference to `depth==1`), but it didn't look any better.